articles/remapping.md

```diff
@@ -0,0 +1,234 @@
+
+**Rationale:**
+ROS 1 has this feature using the environment variable `ROS_NAMESPACE` and the argument `__ns`
+
+- A system has multiple robots
+- A users wants the robots to not interfere with each other, so each robot is put into its own namespace
+- *TODO Global topics like /tf and /tf_static must be individually remapped in ROS 1. What to do about them?*
+
+
+#### A user can remap a name they see on a node after it has launched
+This implies remapping rules are applied in sequence.
+This also implies there is a way to supply remapping rules to a node besides the command line (services?).
+
+**Rationale:**
+A developer may want to make a running node subscribe to a different source of data to see how it behaves.
+It is less work for them to change the name they see rather than digging through the source code to find the original name.
+
+
+### Structure of a Remapping Rule
+Remapping rules have two parts:
+
+1. a part used to determine if the rule applies to a name
+2. a replacement for the names that are matched
+
+
+### Remapping Names in ROS 1
+Remapping is a feature that also exists in ROS 1.
+In ROS 1 remapping works by passing in [arguments](http://wiki.ros.org/Remapping%20Arguments) to each node.
+Client implementations also have APIs in code to pass remapping rules when the node is initialized.
+A remap rule consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+Both sides of a rule are [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b).
+Before a name is remapped it is also [expanded to FQN](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f).
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+
+### Thoughts on ROS 2 implementation
+`rmw` is likely ignorant of remapping.
+The APIs offered by client implementations should remap names automatically to prevent creation of names that are impossible to remap.
+
+
+### Remapping rule command line syntax
+In ROS 1 remapping is done primarily through command line arguments, commonly via roslaunch.
+It may not be possible to use the same syntax for ROS 2 because multiple nodes can be inside of the same process, and it is desired that remap rules are node specific and not process specific.
+This syntax assumes the node's name is guaranteed to be unique in a process.
+It does not support pre-FQN expansion remapping.
+
+- `%`
+ - During matching it is a wildcard that matches a single token
+ - It may appear at most once on the matching side of a rule
+ - During replacement it is replaced with the matched token
+
+- `%%`
+ - During matching it is a wildcard that matches zero or more tokens with at least one slash
+ - It may appear at most once on the matching side of a rule
+ - During replacement it is replaced with the matched slashes and tokens
```

Comment:
I agree that the `|` will be problematic on the command line. I'd suggest that we try to reuse existing syntax. I'd suggest that we consider using the POSIX fnmatch wildcard substitution rules which is already used in the DDS API would be both low cognative load to learn, easy to implement, and no new implementation or dependency. Note that if we want to extend it I believe that all the application specific things like the node specific requirement can simply be done as a preprocessing step before we pass it to fnmatch which implements the wildcard replacements in C.